osvimdriver/service/resourcedriver.py

```diff
@@ -148,6 +168,38 @@ def __handle_create(self, driver_files, system_properties, resource_properties,
         associated_topology = self.__build_associated_topology_response(stack_id)
         return LifecycleExecuteResponse(request_id, associated_topology=associated_topology)
 
+    def __handle_adopt(self, driver_files, system_properties, resource_properties, request_properties, associated_topology, openstack_location):        
+        stack_resource_entry = None
+        if (associated_topology is None or len(associated_topology.to_dict()) != 1):
+            # Expect one adopt stack only per call!
+            raise InvalidRequestError("You must supply exactly one stack_id to adopt in associated_topology")
+        # Only expect one associated_topology so break after first
+        for key, value in associated_topology.to_dict().items():
+            stack_resource_entry = associated_topology.get(str(key))
+            break
+        
+        if stack_resource_entry is None:            
+            # There is no Stack associated to this Resource raise error
+            raise InvalidRequestError("You must supply the stack_id in associated_topology")            
+           
+        stack_id = stack_resource_entry.element_id
+        heat_driver = openstack_location.heat_driver        
+
+        if stack_id != None and len(stack_id.strip())!=0 and stack_id.strip() != "0":
+            try:
+                # Check for valid stack
+                stack_to_adopt = heat_driver.get_stack(stack_id.strip())
+            except StackNotFoundError as e:
+                raise InfrastructureNotFoundError(str(e)) from e
+        # get the status and check it's ok 
```

Comment:
Does the if need an else here in case stack_id is None or doesn't meet the constraints you've imposed? Also, a guard against stack_to_adopt being None?

Reply:
added an else to raise an error